cmd/key_annotate.go

```diff
@@ -2,88 +2,166 @@ package cmd
+			}
+
+			switch rf.Output {
+			case flags.InteractiveOutput:
+				PrintAnnotateKeyResponse(w, req)
+			case flags.JSONOutput:
+				return nil
+			}
 
-	if len(keyAnnotateArgs.metadata) == 0 && !keyAnnotateArgs.clear {
-		return ErrMetaOrClearIsRequired
+			return nil
+		},
 	}
-	if len(keyAnnotateArgs.metadata) != 0 && keyAnnotateArgs.clear {
-		return ErrCanNotHaveBothMetadataAndClearFlagsSet
+
+	cmd.Flags().StringVarP(&f.Wallet,
+		"wallet", "w",
+		"",
+		"Wallet holding the public key",
+	)
+	cmd.Flags().StringVarP(&f.PassphraseFile,
+		"passphrase-file", "p",
+		"",
+		"Path to the file containing the wallet's passphrase",
+	)
+	cmd.Flags().StringVarP(&f.PubKey,
+		"pubkey", "k",
+		"",
+		"Public key to annotate (hex-encoded)",
+	)
+	cmd.Flags().StringSliceVarP(&f.RawMetadata,
+		"meta", "m",
+		[]string{},
+		`A list of metadata e.g: "my-key1:my-value1,my-key2:my-value2"`,
+	)
+	cmd.Flags().BoolVar(&f.Clear,
+		"clear",
+		false,
+		"Clear the metadata",
+	)
+
+	return cmd
+}
+
+type AnnotateKeyFlags struct {
+	Wallet         string
+	PubKey         string
+	PassphraseFile string
+	Clear          bool
+	RawMetadata    []string
+}
+
+func (f *AnnotateKeyFlags) Validate() (*wallet.AnnotateKeyRequest, error) {
+	req := &wallet.AnnotateKeyRequest{}
+
+	if len(f.Wallet) == 0 {
+		return nil, flags.FlagMustBeSpecifiedError("wallet")
 	}
+	req.Wallet = f.Wallet
 
-	passphrase, err := getPassphrase(keyAnnotateArgs.passphraseFile, false)
-	if err != nil {
-		return err
+	if len(f.PubKey) == 0 {
+		return nil, flags.FlagMustBeSpecifiedError("pubkey")
 	}
+	req.PubKey = f.PubKey
 
-	var metadata []wallet.Meta
-	if !keyAnnotateArgs.clear {
-		metadata, err = parseMeta(keyAnnotateArgs.metadata)
-		if err != nil {
-			return err
-		}
+	if len(f.RawMetadata) == 0 && !f.Clear {
+		return nil, flags.OneOfFlagsMustBeSpecifiedError("meta", "clear")
+	}
+	if len(f.RawMetadata) != 0 && f.Clear {
+		return nil, flags.FlagsMutuallyExclusiveError("meta", "clear")
 	}
 
-	err = handler.LoginWallet(keyAnnotateArgs.wallet, passphrase)
+	metadata, err := cli.ParseMetadata(f.RawMetadata)
 	if err != nil {
-		return fmt.Errorf("could not login to the wallet: %w", err)
+		return nil, err
 	}
+	req.Metadata = metadata
 
-	err = handler.UpdateMeta(keyAnnotateArgs.wallet, keyAnnotateArgs.pubkey, passphrase, metadata)
+	passphrase, err := flags.GetPassphrase(f.PassphraseFile)
 	if err != nil {
-		return fmt.Errorf("could not update the metadata: %w", err)
+		return nil, err
 	}
+	req.Passphrase = passphrase
 
-	if rootArgs.output == "human" {
-		p := printer.NewHumanPrinter()
-		if keyAnnotateArgs.clear {
-			p.CheckMark().SuccessText("Annotation cleared").NextLine()
-		} else {
-			p.CheckMark().SuccessText("Annotation succeeded").NextSection()
-			p.Text("New metadata:").NextLine()
-			printMeta(p, metadata)
-		}
+	return req, nil
+}
+
+func PrintAnnotateKeyResponse(w io.Writer, req *wallet.AnnotateKeyRequest) {
+	metadataHaveBeenCleared := len(req.Metadata) != 0
```

Comment:
I think this should be `metadataHaveBeenCleared := len(req.Metadata) == 0`? Locally I'm seeing it print "Annotation cleared." when setting new metadata.

Reply:
Nice catch !